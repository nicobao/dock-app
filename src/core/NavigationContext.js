import React, {useEffect, useState} from 'react';

export const NavigationContext = React.createContext({currentTab: 'tokens'});
export const withNavigationContext = WrappedComponent => props => {
  const [currentTab, setCurrentTab] = useState('tokens');
  const {route} = props;

  useEffect(() => {
    if (route && route?.params?.currentTab) {
      setCurrentTab(route.params.currentTab);
    }
  }, [route]);

  return (
    <NavigationContext.Provider
      value={{
        currentTab,
      }}>
      <WrappedComponent {...props} />
    </NavigationContext.Provider>
  );
};
