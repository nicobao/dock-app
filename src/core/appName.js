export const app = {
  name: 'Dock Wallet',
};
