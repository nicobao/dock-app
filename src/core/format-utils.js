import {
  formatCurrency,
  formatAddress,
  formatDockAmount,
  getPlainDockAmount,
  formatDate,
} from '@docknetwork/wallet-sdk-core/lib/core/format-utils';

export {
  formatCurrency,
  formatAddress,
  formatDockAmount,
  getPlainDockAmount,
  formatDate,
};
