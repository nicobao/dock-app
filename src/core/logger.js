const winston = require('winston');

export const Logger = winston.createLogger({
  format: winston.format.simple(),
  level: 'info',
  transports: [new winston.transports.Console()],
});
