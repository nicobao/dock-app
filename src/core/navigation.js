import {createNavigationContainerRef} from '@react-navigation/native';

export const navigationRef = createNavigationContainerRef();

export function navigate(name, params) {
  if (navigationRef.isReady()) {
    navigationRef.navigate(name, params);
  }
}

export function navigateBack() {
  if (navigationRef.isReady()) {
    navigationRef.goBack();
  }
}
