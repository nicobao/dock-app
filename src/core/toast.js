import {captureException} from '@sentry/react-native';
import {Stack} from 'native-base';
import React from 'react';
import {Pressable} from 'react-native';
import {translate} from 'src/locales';
import {CheckCircleIcon, Text, XCircleIcon} from '../design-system';
import {Theme} from '../design-system/theme';
import {
  ANALYTICS_EVENT,
  logAnalyticsEvent,
} from '../features/analytics/analytics-slice';

let toast;

export function setToast(t) {
  toast = t;
}

const typeMap = {
  success: {
    icon: () => <CheckCircleIcon />,
    bg: '#27272A',
  },
  error: {
    icon: () => <XCircleIcon />,
    bg: Theme.colors.primaryBackground,
  },
  message: {
    icon: () => null,
    bg: Theme.colors.primaryBackground,
  },
};

export const withErrorToast =
  (fn, message) =>
  async (...params) => {
    try {
      return await fn(...params);
    } catch (err) {
      const errorMessage = message || getErrorMessageFromErrorObject(err);
      captureException(err);
      showUnexpectedErrorToast(errorMessage);

      logAnalyticsEvent(ANALYTICS_EVENT.FAILURES, {
        ...params,
        message: errorMessage,
      });
      throw err;
    }
  };

export function showUnexpectedErrorToast(
  message = translate('global.unexpected_error'),
) {
  showToast({
    message,
    type: 'error',
  });
}
export function getErrorMessageFromErrorObject(err) {
  if (typeof err === 'string') {
    if (err.indexOf('AssertionError') > -1) {
      return err.substring(16, err.length - 1).trim();
    }
    return err.trim();
  } else if (err?.message && typeof err.message === 'string') {
    return getErrorMessageFromErrorObject(err.message);
  }
  return translate('global.unexpected_error');
}
export function showToast({message, type = 'success', duration = 2000}) {
  if (!toast) {
    return;
  }

  toast.show({
    duration,
    placement: 'top',
    isClosable: true,
    render: props => {
      const typeProps = typeMap[type] || typeMap.success;

      return (
        <Pressable
          onPress={toast.closeAll}
          _pressed={{
            opacity: Theme.touchOpacity,
          }}>
          <Stack
            bg={typeProps.bg}
            px={6}
            py={4}
            rounded="md"
            mb={5}
            direction="row">
            {typeProps.icon()}
            <Text ml={2} fontWeight={600} fontSize={14}>
              {message}
            </Text>
          </Stack>
        </Pressable>
      );
    },
  });
}
