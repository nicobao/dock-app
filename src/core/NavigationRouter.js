import {NavigationContainer} from '@react-navigation/native';

import React, {useEffect} from 'react';
import {Linking, Platform} from 'react-native';

import {navigationRef} from './navigation';

import DeepLinking from 'react-native-deep-linking';

import {authenticationSelectors} from '../features/unlock-wallet/unlock-wallet-slice';
import {useSelector} from 'react-redux';
import {useDeepLink} from './hooks/deepLinkHooks';
import {UnAuthNavigationStackScreen} from './navigation/unauthenticated-stack';
import {AuthNavigationStackScreen} from './navigation/authenticated-stack';

const forFade = ({current, closing}) => ({
  cardStyle: {
    opacity: current.progress,
  },
});

const screenOptions = {
  headerShown: false,
};

if (Platform.OS === 'android') {
  screenOptions.cardStyleInterpolator = forFade;
}

export function NavigationRouter() {
  const isLoggedIn = useSelector(authenticationSelectors.isLoggedIn);
  const {gotoScreenDeepLink} = useDeepLink({isLoggedIn});

  useEffect(() => {
    const getAsyncURL = async () => {
      const initialUrl = await Linking.getInitialURL();
      if (initialUrl !== undefined && initialUrl != null) {
        gotoScreenDeepLink({url: initialUrl});
      }
    };

    getAsyncURL();
  }, [gotoScreenDeepLink]);

  useEffect(() => {
    DeepLinking.addScheme('dockwallet://');
    Linking.addEventListener('url', gotoScreenDeepLink);

    return () => {
      Linking.removeEventListener('url', gotoScreenDeepLink);
    };
  }, [gotoScreenDeepLink]);

  return (
    <NavigationContainer ref={navigationRef}>
      {isLoggedIn ? (
        <AuthNavigationStackScreen />
      ) : (
        <UnAuthNavigationStackScreen />
      )}
    </NavigationContainer>
  );
}
