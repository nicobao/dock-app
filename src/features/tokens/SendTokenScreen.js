import {useAccount} from '@docknetwork/wallet-sdk-react-native/lib';
import Clipboard from '@react-native-community/clipboard';
import {Box, FormControl, Input, Stack} from 'native-base';
import React, {useEffect, useState} from 'react';
import {Keyboard} from 'react-native';
import Share from 'react-native-share';
import {useDispatch} from 'react-redux';
import {NumericKeyboard} from 'src/components/NumericKeyboard';
import {PolkadotIcon} from 'src/components/PolkadotIcon';
import {navigate} from 'src/core/navigation';

import {utilCryptoService} from '@docknetwork/wallet-sdk-core/lib/services/util-crypto';
import BigNumber from 'bignumber.js';
import {Routes} from 'src/core/routes';
import {showToast} from 'src/core/toast';

import {formatCurrency} from 'src/core/format-utils';
import {
  BackButton,
  Button,
  Content,
  Header,
  LoadingButton,
  ScreenContainer,
  Typography,
} from '../../design-system';
import {translate} from '../../locales';
import {transactionsOperations} from '../transactions/transactions-slice';
import {ConfirmTransactionModal, TokenAmount} from './ConfirmTransactionModal';

export function SendTokenScreen({form, onChange, onScanQRCode, onNext}) {
  const onChangeAddressMethod = onChange('recipientAddress');

  async function onChangeAddress(v) {
    onChangeAddressMethod(v);
    if (v) {
      const copiedContent = await Clipboard.getString();
      if (copiedContent && v.includes(copiedContent)) {
        Keyboard.dismiss();
      }
    }
  }

  return (
    <ScreenContainer testID="unlockWalletScreen">
      <Header>
        <BackButton />
      </Header>
      <Content marginLeft={26} marginRight={26}>
        <Stack>
          <Typography variant="h1" mb={2}>
            {translate('send_token.title')}
          </Typography>
        </Stack>
        <Box my={7}>
          <Stack>
            <FormControl isInvalid={form._errors.recipientAddress}>
              <Input
                value={form.recipientAddress}
                onChangeText={onChangeAddress}
                autoCapitalize="none"
                placeholder={translate('send_token.recipient_address')}
                pasteFromClipboard
              />
              <FormControl.ErrorMessage>
                {form._errors.recipientAddress}
              </FormControl.ErrorMessage>
            </FormControl>
          </Stack>
          <Stack direction="row" alignItems="center" mt={3}>
            <Button colorScheme="secondary" size="sm" onPress={onScanQRCode}>
              {translate('send_token.scan_qr_code')}
            </Button>
          </Stack>
        </Box>
        <Stack flex={1} alignContent="flex-end">
          <LoadingButton onPress={onNext} mb={4}>
            {translate('navigation.next')}
          </LoadingButton>
        </Stack>
      </Content>
    </ScreenContainer>
  );
}

export function EnterTokenAmount({form, onMax, onChange, onBack, onNext}) {
  return (
    <ScreenContainer testID="unlockWalletScreen">
      <Header>
        <BackButton onPress={onBack} />
      </Header>
      <Content marginLeft={26} marginRight={26}>
        <Stack>
          <Stack direction="row" alignItems="center" justifyContent="center">
            <Typography
              variant="h1"
              style={{fontSize: 48, lineHeight: 60}}
              mr={2}>{`${form.amount || 0}`}</Typography>
            <Typography variant="h1">{form.tokenSymbol}</Typography>
          </Stack>
          <TokenAmount amount={form.amount || 0} symbol={form.tokenSymbol}>
            {({fiatSymbol, fiatAmount}) => (
              <Stack direction="row" justifyContent="center">
                <Typography>{formatCurrency(fiatAmount)} </Typography>
                <Typography>{fiatSymbol}</Typography>
              </Stack>
            )}
          </TokenAmount>
          <FormControl isInvalid={form._errors.amount}>
            <FormControl.ErrorMessage>
              {form._errors.amount}
            </FormControl.ErrorMessage>
          </FormControl>
        </Stack>
        <Stack>
          <Button colorScheme="secondary" mb={2} onPress={onMax}>
            {translate('send_token.send_max')}
          </Button>
        </Stack>
        <Box my={7}>
          <NumericKeyboard
            allowDecimal
            onChange={onChange('amount')}
            value={form.amount}
          />
        </Box>
        <Stack flex={1} alignContent="flex-end">
          <Button onPress={onNext} mb={4}>
            {translate('navigation.next')}
          </Button>
        </Stack>
      </Content>
    </ScreenContainer>
  );
}

const Steps = {
  sendTo: 1,
  enterAmount: 2,
};

export function handleFeeUpdate({
  updateForm,
  account,
  form,
  fee,
  setShowConfirmation,
}) {
  const accountBalance = account.balance;
  const amountAndFees = BigNumber(form.amount).plus(fee);

  if (fee >= accountBalance) {
    showToast({
      message: translate('send_token.insufficient_balance'),
      type: 'error',
    });
    return false;
  }

  const formUpdates = {
    amountMessage: null,
  };

  if (amountAndFees > accountBalance) {
    const newAmount = BigNumber(account.balance).minus(fee).toNumber();

    formUpdates.amount = newAmount;

    if (!form.sendMax) {
      formUpdates.amountMessage = translate('send_token.amount_minus_fees_msg');
    }
  }

  updateForm({
    ...formUpdates,
    fee,
  });

  setShowConfirmation(true);

  return true;
}

const defaultFormState = {
  recipientAddress: '',
  amount: 0,
  tokenSymbol: 'DOCK',
  fee: 0,
  validating: false,
  _errors: {},
  _hasError: false,
};

export function SendTokenContainer({route}) {
  const dispatch = useDispatch();
  const {address, recipientAddress} = route.params || {};
  const {account} = useAccount(address);

  const [showConfirmation, setShowConfirmation] = useState();
  const [step, setStep] = useState(Steps.sendTo);
  const [form, setForm] = useState({
    ...defaultFormState,
    recipientAddress,
  });

  const handleCopyAddress = () => {
    Clipboard.setString(address);
    showToast({
      message: translate('receive_token.address_copied'),
    });
  };

  const handleShareAddress = () =>
    Share.open({
      message: address,
    });

  const handleChange = key => evt => {
    if (form[key] === evt) {
      return;
    }

    updateForm({
      sendMax: false,
      [key]: evt,
    });
  };

  const updateForm = newValues =>
    setForm(v => ({
      ...v,
      ...newValues,
    }));

  useEffect(() => {
    const toAddress = route.params.recipientAddress;
    if (toAddress && toAddress !== form.recipientAddress) {
      updateForm({
        recipientAddress: toAddress,
      });
    }
  }, [route.params, form.recipientAddress]);

  if (step === Steps.sendTo) {
    return (
      <SendTokenScreen
        form={form}
        onCopyAddress={handleCopyAddress}
        onScanQRCode={() => {
          navigate(
            Routes.APP_QR_SCANNER,
            {
              onData: async toAddress => {
                const addressValid = await utilCryptoService.isAddressValid(
                  toAddress,
                );

                if (addressValid) {
                  navigate(
                    Routes.TOKEN_SEND,
                    {
                      address,
                      recipientAddress: toAddress,
                    },
                    true,
                  );
                } else {
                  navigate(Routes.TOKEN_SEND, route.params);
                  showToast({
                    message: translate('global.not_valid_address'),
                    type: 'error',
                  });
                }
              },
            },
            true,
          );
        }}
        onShareAddress={handleShareAddress}
        onChange={handleChange}
        onNext={async () => {
          if (form.recipientAddress === account.address) {
            return setForm(v => ({
              ...v,
              _errors: {
                recipientAddress: translate('send_token.same_address'),
              },
            }));
          }

          const addressValid = await utilCryptoService.isAddressValid(
            form.recipientAddress,
          );

          if (!addressValid) {
            return setForm(v => ({
              ...v,
              _errors: {
                recipientAddress: translate('send_token.invalid_address'),
              },
            }));
          }

          setStep(Steps.enterAmount);
        }}
      />
    );
  }

  if (step === Steps.enterAmount) {
    return (
      <>
        <ConfirmTransactionModal
          onClose={() => {
            setShowConfirmation(false);
          }}
          onConfirm={() => {
            dispatch(
              transactionsOperations.sendTransaction({
                ...form,
                accountAddress: account.address,
              }),
            ).finally(() => {
              setShowConfirmation(false);
              navigate(Routes.ACCOUNT_DETAILS, {
                id: account.address,
              });
              setForm(defaultFormState);
              setStep(Steps.sendTo);
            });
          }}
          amountMessage={form.amountMessage}
          visible={showConfirmation}
          accountIcon={<PolkadotIcon address={form.recipientAddress} />}
          tokenSymbol={form.tokenSymbol}
          sentAmount={form.amount}
          fee={form.fee}
          recipientAddress={form.recipientAddress}
        />
        <EnterTokenAmount
          form={form}
          onChange={handleChange}
          onNext={() => {
            let amountError;

            if (form.amount <= 0) {
              amountError = translate('send_token.invalid_amount');
            } else if (form.amount > account.balance) {
              amountError = translate('send_token.insufficient_funds');
            }

            if (amountError) {
              return setForm(v => ({
                ...v,
                _errors: {
                  amount: amountError,
                },
              }));
            }

            return dispatch(
              transactionsOperations.getFeeAmount({
                ...form,
                accountAddress: account.address,
              }),
            ).then(fee => {
              return handleFeeUpdate({
                account,
                form,
                fee,
                updateForm,
                setShowConfirmation,
              });
            });
          }}
          onBack={() => {
            setStep(Steps.sendTo);
          }}
          onMax={() => {
            updateForm({
              amount: account.balance,
              sendMax: true,
            });
          }}
          tokenSymbol="DOCK"
        />
      </>
    );
  }

  return null;
}
