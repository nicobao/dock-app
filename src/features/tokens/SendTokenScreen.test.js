import React from 'react';
import {shallow} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import {SendTokenScreen, handleFeeUpdate} from './SendTokenScreen';
import {setToast} from '../../core/toast';
import {getPlainDockAmount} from 'src/core/format-utils';

const mockStore = configureMockStore();

describe('SendTokenScreen', () => {
  it('should render correctly', () => {
    const initialState = {
      // placeholder for redux store
    };

    const wrapper = shallow(
      <SendTokenScreen
        address={'3B9W5fS4ygoUVnGhzTk7sihSdhCyJQRNMQ7aJfDdsfWb16f1'}
        form={{
          _errors: {},
        }}
        onChange={() => {}}
      />,
      {
        context: {store: mockStore(initialState)},
      },
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  describe('handle fee update', () => {
    beforeEach(() => {
      const toastMock = {
        show: jest.fn(),
      };

      setToast(toastMock);
    });

    it('enough balance to pay for fees', () => {
      const setShowConfirmation = jest.fn();
      let form = {
        amount: 1,
      };
      const updateForm = f => {
        form = f;
      };
      const balance = getPlainDockAmount(100);
      const fee = 2;

      const result = handleFeeUpdate({
        account: {
          balance,
        },
        fee,
        form,
        setShowConfirmation,
        updateForm,
      });

      expect(result).toBe(true);
      expect(form.fee).toBe(fee);
      expect(setShowConfirmation).toBeCalledWith(true);
    });

    it('expect to reduce the amount to pay for fees', () => {
      const setShowConfirmation = jest.fn();
      let form = {
        amount: 1,
      };
      const updateForm = f => {
        form = {
          ...form,
          ...f,
        };
      };
      const balance = 3;
      const fee = 2;

      const result = handleFeeUpdate({
        account: {
          balance,
        },
        fee,
        form,
        setShowConfirmation,
        updateForm,
      });

      expect(result).toBe(true);
      expect(form.fee).toBe(fee);
      expect(form.amountMessage).toBeDefined();
      expect(form.amount).toBe(1);
      expect(setShowConfirmation).toBeCalledWith(true);
    });

    it('expect to not have enough balance for fees', () => {
      const setShowConfirmation = jest.fn();
      let form = {
        amount: 2,
      };
      const updateForm = f => {
        form = f;
      };
      const balance = getPlainDockAmount(1);
      const fee = getPlainDockAmount(2);

      const result = handleFeeUpdate({
        account: {
          balance,
        },
        fee,
        form,
        setShowConfirmation,
        updateForm,
      });

      expect(result).toBe(false);
      expect(form).toBe(form);
      expect(setShowConfirmation).not.toBeCalled();
    });
  });
});
