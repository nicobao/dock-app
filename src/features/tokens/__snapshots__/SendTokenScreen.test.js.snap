// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendTokenScreen should render correctly 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "backgroundColor": "#141417",
      "flex": 1,
    }
  }
  testID="unlockWalletScreen"
>
  <AppGlobalHeader />
  <Memo(ForwardRef(Box))
    flex={1}
  >
    <Memo(ForwardRef(Box))
      flex={1}
    >
      <Styled(View)>
        <BackButton />
      </Styled(View)>
      <Styled(ScrollView)
        marginLeft={26}
        marginRight={26}
      >
        <Memo(ForwardRef(Stack))>
          <Styled(Component)
            mb={2}
            variant="h1"
          >
            Send to
          </Styled(Component)>
        </Memo(ForwardRef(Stack))>
        <Memo(ForwardRef(Box))
          my={7}
        >
          <Memo(ForwardRef(Stack))>
            <Memo(ForwardRef(FormControl))>
              <Memo(ForwardRef(Input))
                autoCapitalize="none"
                onChangeText={[Function]}
                pasteFromClipboard={true}
                placeholder="Recipient address"
              />
              <Memo(ForwardRef(FormControlErrorMessage)) />
            </Memo(ForwardRef(FormControl))>
          </Memo(ForwardRef(Stack))>
          <Memo(ForwardRef(Stack))
            alignItems="center"
            direction="row"
            mt={3}
          >
            <Button
              colorScheme="secondary"
              size="sm"
            >
              Scan QR Code
            </Button>
          </Memo(ForwardRef(Stack))>
        </Memo(ForwardRef(Box))>
        <Memo(ForwardRef(Stack))
          alignContent="flex-end"
          flex={1}
        >
          <LoadingButton
            mb={4}
          >
            Next
          </LoadingButton>
        </Memo(ForwardRef(Stack))>
      </Styled(ScrollView)>
    </Memo(ForwardRef(Box))>
  </Memo(ForwardRef(Box))>
</RNCSafeAreaView>
`;
